core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -277,7 +311,8 @@ protected PlanNodeStatsEstimate visitIsNullPredicate(IsNullPredicate node, Void
         @Override
         protected PlanNodeStatsEstimate visitBetweenPredicate(BetweenPredicate node, Void context)
         {
-            if (!(node.getValue() instanceof SymbolReference)) {
+            SymbolStatsEstimate valueStats = getExpressionStats(node.getValue());
```

Comment:
`FilterStatsCalculator` gains special logic to ignore coercions and thus produce better stats. Here, however, we're delegating to `scalarStatsCalculator`, which has different logic.

- in getExpressionStats, if `getSymbolReferenceForStats().isPresent()`, we should use this symbols stats
- in `scalarStatsCalculator` we should fix visitCast to honor `areStatsRepresentationsCompatible`